src/core/models.py

```diff
@@ -20,26 +28,46 @@ class StateBase:
     """Base state object."""
 
     def __init__(
-        self, relation: Relation | None, component: Unit | Application, substrate: SUBSTRATES
+        self,
+        relation: Relation | None,
+        data_interface: DataRelation,
+        component: Unit | Application,
+        substrate: SUBSTRATES,
     ):
         self.relation = relation
+        self.data_interface = data_interface
         self.component = component
         self.substrate = substrate
 
     @property
     def relation_data(self) -> MutableMapping[str, str]:
         """The raw relation data."""
-        if not self.relation:
+        if not self.relation or not self.data_interface:
             return {}
 
-        return self.relation.data[self.component]
+        my_data = {}
+        if my_data_dict := self.data_interface.fetch_my_relation_data([self.relation.id]):
+            my_data = my_data_dict.get(self.relation.id, {})
+
+        try:
+            other_data = self.data_interface.fetch_relation_data([self.relation.id])[
+                self.relation.id
+            ]
+        except NotImplementedError:
+            other_data = {}
+        return my_data | other_data
 
     def update(self, items: dict[str, str]) -> None:
         """Writes to relation_data."""
-        if not self.relation:
+        if not self.relation or not self.data_interface:
             return
 
-        self.relation_data.update(items)
+        delete_fields = [key for key in items if not items[key]]
```

Comment:
**question:** Maybe I'm being dim, but wouldn't this line ALWAYS return an empty list?

You're looping over a dict, which loops over keys. You're then keeping keys, that aren't keys in the dict, which they all are, because that's the loop value.

Reply:
This is the point:
```
In [4]: items = {'a': '', 'b': "something"}

In [5]: [key for key in items if not items[key]]
Out[5]: ['a']
```

Using `update()` (but it's the same for `set_secret()` in old-style charms) the way we delete is to `update()` with `""`.

That's exactly what this list comprehension is distinguishing for.

Reply:
Ah, right got it.

May I suggest making that a tad more explicit?:

```python
[for key:value in items.items() if value]
```